File: backend/src/main/java/com/explorateur/backend/service/JournalService.java
package com.explorateur.backend.service;

import com.explorateur.backend.dto.JournalFilterRequest;
import com.explorateur.backend.dto.JournalResponse;
import com.explorateur.backend.entity.Journal;
import com.explorateur.backend.entity.Utilisateur;
import com.explorateur.backend.repository.JournalRepository;
import com.explorateur.backend.repository.UtilisateurRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service pour la gestion du journal d'audit
 */
@Service
@RequiredArgsConstructor
public class JournalService {

    private final JournalRepository journalRepository;
    private final UtilisateurRepository utilisateurRepository;

    /**
     * Enregistre une action dans le journal
     */
    @Transactional
    public void logAction(String action) {
        try {
            Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
            
            if (authentication != null && authentication.isAuthenticated() 
                && !"anonymousUser".equals(authentication.getPrincipal())) {
                
                String username = authentication.getName();
                Utilisateur utilisateur = utilisateurRepository.findByUsername(username)
                        .orElse(null);

                Journal journal = Journal.builder()
                        .action(action)
                        .utilisateur(utilisateur)
                        .timestamp(LocalDateTime.now())
                        .build();

                journalRepository.save(journal);
            }
        } catch (Exception e) {
            // Ne pas propager les erreurs de journalisation
            System.err.println("Erreur lors de la journalisation: " + e.getMessage());
        }
    }

    /**
     * Enregistre une action avec un utilisateur spécifique
     */
    @Transactional
    public void logAction(String action, Long utilisateurId) {
        try {
            Utilisateur utilisateur = null;
            if (utilisateurId != null) {
                utilisateur = utilisateurRepository.findById(utilisateurId).orElse(null);
            }

            Journal journal = Journal.builder()
                    .action(action)
                    .utilisateur(utilisateur)
                    .timestamp(LocalDateTime.now())
                    .build();

            journalRepository.save(journal);
        } catch (Exception e) {
            System.err.println("Erreur lors de la journalisation: " + e.getMessage());
        }
    }

    /**
     * Récupère toutes les entrées du journal
     */
    @Transactional(readOnly = true)
    public List<JournalResponse> getAllJournal() {
        return journalRepository.findAllByOrderByTimestampDesc().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Récupère les entrées du journal avec filtres
     */
    @Transactional(readOnly = true)
    public List<JournalResponse> getJournalWithFilters(JournalFilterRequest filter) {
        List<Journal> journals = journalRepository.findByFilters(
                filter.getDateDebut(),
                filter.getDateFin(),
                filter.getUtilisateurId(),
                filter.getSearchText()
        );

        return journals.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Récupère les entrées du journal par période
     */
    @Transactional(readOnly = true)
    public List<JournalResponse> getJournalByPeriod(LocalDateTime dateDebut, LocalDateTime dateFin) {
        return journalRepository.findByTimestampBetweenOrderByTimestampDesc(dateDebut, dateFin)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Récupère les entrées du journal par utilisateur
     */
    @Transactional(readOnly = true)
    public List<JournalResponse> getJournalByUtilisateur(Long utilisateurId) {
        return journalRepository.findByUtilisateurIdOrderByTimestampDesc(utilisateurId)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Mapper Journal vers JournalResponse
     */
    private JournalResponse mapToResponse(Journal journal) {
        return JournalResponse.builder()
                .id(journal.getId())
                .action(journal.getAction())
                .username(journal.getUtilisateur() != null ? journal.getUtilisateur().getUsername() : "Système")
                .utilisateurId(journal.getUtilisateur() != null ? journal.getUtilisateur().getId() : null)
                .timestamp(journal.getTimestamp())
                .build();
    }
}

File: frontend/src/services/journal.service.ts
import apiService from './api.service';
import type { JournalEntry, JournalFilterRequest } from '../types';

/**
 * Service pour la gestion du journal d'audit
 */
class JournalService {
  private readonly baseUrl = '/api/journal';
  private readonly queryUrl = '/api/journal-query/entries';

  /**
   * Récupérer toutes les entrées du journal
   */
  async getAllJournal(): Promise<JournalEntry[]> {
    return apiService.get<JournalEntry[]>(this.baseUrl);
  }

  /**
   * Filtrer les entrées du journal avec critères multiples
   */
  async filterJournal(filter: JournalFilterRequest): Promise<JournalEntry[]> {
    const params = new URLSearchParams();

    if (filter.dateDebut) {
      params.append('dateDebut', filter.dateDebut);
    }
    if (filter.dateFin) {
      params.append('dateFin', filter.dateFin);
    }
    if (filter.utilisateurId) {
      params.append('utilisateurId', filter.utilisateurId.toString());
    }
    if (filter.searchText) {
      params.append('searchText', filter.searchText);
    }

    const query = params.toString();
    const url = query ? `${this.queryUrl}?${query}` : this.queryUrl;
    try {
      return await apiService.get<JournalEntry[]>(url);
    } catch (error: any) {
      if (error?.response?.status !== 403) {
        throw error;
      }

      const entries = await this.getAllJournal();
      return entries.filter((entry) => {
        const entryDate = new Date(entry.timestamp);
        const afterStart = filter.dateDebut ? entryDate >= new Date(filter.dateDebut) : true;
        const beforeEnd = filter.dateFin ? entryDate <= new Date(filter.dateFin) : true;
        const byUser = filter.utilisateurId ? entry.utilisateurId === filter.utilisateurId : true;
        const byText = filter.searchText
          ? entry.action.toLowerCase().includes(filter.searchText.toLowerCase()) ||
            entry.username.toLowerCase().includes(filter.searchText.toLowerCase())
          : true;

        return afterStart && beforeEnd && byUser && byText;
      });
    }
  }

  /**
   * Récupérer les entrées par période
   */
  async getJournalByPeriod(dateDebut: string, dateFin: string): Promise<JournalEntry[]> {
    return apiService.get<JournalEntry[]>(
      `${this.baseUrl}/period?dateDebut=${dateDebut}&dateFin=${dateFin}`
    );
  }

  /**
   * Récupérer les entrées par utilisateur
   */
  async getJournalByUtilisateur(utilisateurId: number): Promise<JournalEntry[]> {
    return apiService.get<JournalEntry[]>(`${this.baseUrl}/utilisateur/${utilisateurId}`);
  }
}

export default new JournalService();
